src/main/java/com/ibm/stocator/fs/swift/SwiftAPIClient.java

```diff
@@ -311,39 +311,40 @@ an request on an object (not a container) that has a trailing slash will lead
       */
       objectNameNoSlash = objectName.substring(0, objectName.length() - 1);
     }
-    StoredObject so = cont.getObject(objectNameNoSlash);
+
+    HttpResponse metadata = SwiftAPIDirect.getObjectMetadata(mJossAccount, container,
+            objectNameNoSlash, swiftConnectionManager);
+
     boolean isDirectory = false;
-    if (so.exists()) {
-      // We need to check if the object size is equal to zero
-      // If so, it might be a directory
-      long contentLength = so.getContentLength();
-      String lastModified = so.getLastModified();
-      if (contentLength == 0) {
-        Collection<DirectoryOrObject> directoryFiles = cont.listDirectory(objectName, '/', "", 10);
-        if (directoryFiles != null && directoryFiles.size() != 0) {
-          // The zero length object is a directory
+
+    if (metadata != null) {
+
+      int contentLength = new Integer(metadata.getFirstHeader("Content-Length").getValue());
+      String contentType = metadata.getFirstHeader("Content-Type").getValue();
+
+      long lastModified = 0L;
+      if (metadata.containsHeader("Last-Modified")) {
+        lastModified = getLastModified(metadata.getFirstHeader("Last-Modified").getValue());
+      }
+
+      if (contentLength == 0) { // Check if directory
+        if (contentType.equals("application/directory")) {
           isDirectory = true;
+        } else if (!contentType.equals("application/octet-stream")) {
+          return null;
         }
+
       }
       LOG.trace("{} is object. isDirectory: {}  lastModified: {}", path.toString(),
-          isDirectory, lastModified);
+              isDirectory, lastModified);
       return new FileStatus(contentLength, isDirectory, 1, blockSize,
-              getLastModified(lastModified), path);
-    }
-    // We need to check if it may be a directory with no zero byte file associated
-    LOG.trace("Checking if directory without 0 byte object associated {}", objectName);
-    Collection<DirectoryOrObject> directoryFiles = cont.listDirectory(objectName + "/", '/',
-        "", 10);
-    if (directoryFiles != null) {
-      LOG.trace("{} got {} candidates", objectName + "/", directoryFiles.size());
-    }
-    if (directoryFiles != null && directoryFiles.size() != 0) {
-      // In this case there is no lastModified
-      isDirectory = true;
-      LOG.debug("Got object {}. isDirectory: {}  lastModified: {}", path, isDirectory, null);
-      return new FileStatus(0, isDirectory, 1, blockSize, 0L, path);
+              lastModified, path);
```

Comment:
@gilv I added this logic back. If the SwiftDirectAPI returns a FileNotFoundException then we check if the path given is a prefix to an object.